Bug: JSON serialization crash with __SwiftValue types in GraphQLMetadataExtractor

## Description

`GraphQLMetadataExtractor.extractVariables()` crashes when Apollo GraphQL variables contain `__SwiftValue` types.

## Error

```
Invalid type in JSON write (__SwiftValue) (NSInvalidArgumentException)
```

**Location:** `GraphQLMetadataExtractor.swift:28`

## Environment

- Apollo iOS: 1.24.0
- dd-sdk-ios-apollo-interceptor: develop branch (commit f3f2b46)

## Root Cause

The code attempts to serialize Apollo variables using `JSONSerialization`:

```swift
let jsonEncodableDict = variables._jsonEncodableObject
let jsonData = try JSONSerialization.data(withJSONObject: jsonEncodableDict, options: [])
```

When `jsonEncodableDict` contains `__SwiftValue` types (which aren't JSON-serializable), `JSONSerialization` throws an `NSInvalidArgumentException`. This is an Objective-C exception, not a Swift Error, so it cannot be caught by the surrounding `do-catch` block.

## Reproduction

1. Create a GraphQL mutation with a complex input type containing nested Swift objects
2. Execute the mutation through Apollo client with `DatadogApolloInterceptor` in the interceptor chain
3. The interceptor will crash when trying to serialize the variables dictionary

Example:
```swift
// GraphQL operation with complex nested input
let mutation = MyMutation(input: MyInput(
    customObject: SomeSwiftType() // Gets wrapped as __SwiftValue
))

// Using ApolloClient with DatadogApolloInterceptor
client.perform(mutation: mutation) // Crashes in extractVariables()
```

## Impact

The uncaught exception causes app crashes when using GraphQL operations with complex Swift types as variables.

Here's a failing test against your codebase.

```diff
diff --git a/Tests/DatadogApolloTests/DatadogApolloTests.swift b/Tests/DatadogApolloTests/DatadogApolloTests.swift
index b1a9f77..e4ea644 100644
--- a/Tests/DatadogApolloTests/DatadogApolloTests.swift
+++ b/Tests/DatadogApolloTests/DatadogApolloTests.swift
@@ -109,4 +109,28 @@ public final class DatadogApolloInterceptorTests: XCTestCase {
         XCTAssertTrue(payload.contains("query GetUser"))
         XCTAssertTrue(payload.contains("user(id: $userId)"))
     }
+
+    // MARK: - __SwiftValue Crash Tests (Issue #17)
+
+    /// This test demonstrates the crash that occurs when JSONSerialization encounters __SwiftValue
+    /// Run this test to see the actual NSInvalidArgumentException that cannot be caught by do-catch
+    func testJSONSerializationCrashesWithSwiftValue() throws {
+        // Create a dictionary that will contain a __SwiftValue wrapper
+        // This happens when you try to use a custom Swift class as a dictionary value
+        let customObject = NSObject() // Any non-JSON-serializable object
+        let problematicDict: [String: Any] = [
+            "input": customObject
+        ]
+
+        // This will crash with: "Invalid type in JSON write (__SwiftValue) (NSInvalidArgumentException)"
+        // The do-catch cannot catch this because it's an Objective-C exception, not a Swift Error
+        do {
+            let _ = try JSONSerialization.data(withJSONObject: problematicDict, options: [])
+            XCTFail("Expected JSONSerialization to throw, but it succeeded")
+        } catch {
+            // This catch block will NEVER be reached because NSInvalidArgumentException
+            // is an Objective-C exception that bypasses Swift's error handling
+            XCTFail("Caught error: \(error), but we expected an uncatchable NSException crash")
+        }
+    }
 }
```

```shell
swift test
Building for debugging...
[2/2] Write swift-version--58304C5D6DBC2206.txt
Build complete! (0.08s)
Test Suite 'All tests' started at 2025-11-19 16:43:10.725.
Test Suite 'DatadogApolloPackageTests.xctest' started at 2025-11-19 16:43:10.725.
Test Suite 'DatadogApolloInterceptorTests' started at 2025-11-19 16:43:10.725.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testCreateInterceptor]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testCreateInterceptor]' passed (0.001 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractOperationTypeFromOperations]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractOperationTypeFromOperations]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractPayloadContainsQueryDocument]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractPayloadContainsQueryDocument]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractPayloadWithEmptyOperationName]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractPayloadWithEmptyOperationName]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractVariablesFromOperationWithNoVariables]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractVariablesFromOperationWithNoVariables]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractVariablesFromQueryOperation]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testExtractVariablesFromQueryOperation]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testGraphQLHeadersConstants]' started.
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testGraphQLHeadersConstants]' passed (0.000 seconds).
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testJSONSerializationCrashesWithSwiftValue]' started.
/Users/blimmer/code/homebound/dd-sdk-ios-apollo-interceptor/Tests/DatadogApolloTests/DatadogApolloTests.swift:128: error: -[DatadogApolloTests.DatadogApolloInterceptorTests testJSONSerializationCrashesWithSwiftValue] : Invalid type in JSON write (NSObject) (NSInvalidArgumentException)
Test Case '-[DatadogApolloTests.DatadogApolloInterceptorTests testJSONSerializationCrashesWithSwiftValue]' failed (0.020 seconds).
Test Suite 'DatadogApolloInterceptorTests' failed at 2025-11-19 16:43:10.748.
	 Executed 8 tests, with 1 failure (1 unexpected) in 0.022 (0.022) seconds
Test Suite 'DatadogApolloPackageTests.xctest' failed at 2025-11-19 16:43:10.748.
	 Executed 8 tests, with 1 failure (1 unexpected) in 0.022 (0.022) seconds
Test Suite 'All tests' failed at 2025-11-19 16:43:10.748.
	 Executed 8 tests, with 1 failure (1 unexpected) in 0.022 (0.023) seconds
◇ Test run started.
↳ Testing Library Version: 1400
↳ Target Platform: arm64e-apple-macos14.0
✔ Test run with 0 tests in 0 suites passed after 0.001 seconds.
make: *** [test] Error 1
```

--- a/Sources/DatadogApollo/Internal/GraphQLMetadataExtractor.swift
+++ b/Sources/DatadogApollo/Internal/GraphQLMetadataExtractor.swift
@@ -22,10 +22,20 @@ internal struct GraphQLMetadataExtractor {
             return nil
         }
 
-        // Use Apollo's built-in JSONEncodable conversion to handle proper serialization
-        let jsonEncodableDict = variables._jsonEncodableObject
+        // Convert each variable to its _jsonValue representation to ensure JSON-serializability.
+        // We use _jsonValue instead of _jsonEncodableObject to properly handle types like
+        // GraphQLEnum that need conversion before serialization.
+        let jsonValueDict = variables.mapValues { variableValue -> Any in
+            // Use _jsonEncodableValue to get the JSONEncodable, then extract _jsonValue
+            if let encodable = variableValue._jsonEncodableValue {
+                return encodable._jsonValue
+            }
+            // Fallback to the variable value itself if it doesn't have an encodable value
+            return variableValue
+        }
+
         do {
-            let jsonData = try JSONSerialization.data(withJSONObject: jsonEncodableDict, options: [])
+            let jsonData = try JSONSerialization.data(withJSONObject: jsonValueDict, options: [])
             return String(data: jsonData, encoding: .utf8)
         } catch {
             // If serialization fails, return nil to avoid breaking the request

--- a/Sources/DatadogApollo/Internal/GraphQLMetadataExtractor.swift
+++ b/Sources/DatadogApollo/Internal/GraphQLMetadataExtractor.swift
@@ -27,11 +27,8 @@ internal struct GraphQLMetadataExtractor {
         // GraphQLEnum that need conversion before serialization.
         let jsonValueDict = variables.mapValues { variableValue -> Any in
             // Use _jsonEncodableValue to get the JSONEncodable, then extract _jsonValue
-            if let encodable = variableValue._jsonEncodableValue {
-                return encodable._jsonValue
-            }
             // Fallback to the variable value itself if it doesn't have an encodable value
-            return variableValue
+            return variableValue._jsonEncodableValue?._jsonValue ?? variableValue
         }
 
         do {

--- a/Tests/DatadogApolloTests/DatadogApolloTests.swift
+++ b/Tests/DatadogApolloTests/DatadogApolloTests.swift
@@ -109,4 +109,24 @@ public final class DatadogApolloInterceptorTests: XCTestCase {
         XCTAssertTrue(payload.contains("query GetUser"))
         XCTAssertTrue(payload.contains("user(id: $userId)"))
     }
+
+    // MARK: - GraphQL Enum Variables Tests
+
+    func testExtractVariablesWithGraphQLEnum() throws {
+        // Test that GraphQLEnum types are properly serialized to JSON
+        let extractor = GraphQLMetadataExtractor()
+        let operation = MockOperationWithGraphQLEnum(id: "test-123", enumValue: .optionB)
+
+        let result = extractor.extractVariables(from: operation)
+
+        XCTAssertNotNil(result)
+        guard let variables = result else {
+            XCTFail("Expected non-nil variables")
+            return
+        }
+
+        // Verify the enum is serialized as its string value
+        XCTAssertTrue(variables.contains("OPTION_B"))
+        XCTAssertTrue(variables.contains("test-123"))
+    }
 }

--- a/Tests/DatadogApolloTests/TestMocks.swift
+++ b/Tests/DatadogApolloTests/TestMocks.swift
@@ -130,3 +130,32 @@ internal class MockNoVariablesOperation: GraphQLOperation {
 
     var __variables: [String: GraphQLOperationVariableValue]? = nil
 }
+
+// MARK: - Mock GraphQL Enum for Testing
+
+/// Mock enum for testing GraphQLEnum serialization
+internal enum MockEnumValue: String, Sendable, EnumType {
+    case optionA = "OPTION_A"
+    case optionB = "OPTION_B"
+    case optionC = "OPTION_C"
+}
+
+/// Mock operation with GraphQLEnum variables for testing serialization
+internal class MockOperationWithGraphQLEnum: GraphQLOperation {
+    typealias Data = MockData
+
+    static let operationName: String = "QueryWithEnum"
+    static let operationType: GraphQLOperationType = .query
+    static let operationDocument: ApolloAPI.OperationDocument = .init(
+        definition: .init("query QueryWithEnum($id: ID!, $enumValue: EnumValue!) { item(id: $id, enumValue: $enumValue) { id } }")
+    )
+
+    var __variables: [String: GraphQLOperationVariableValue]?
+
+    init(id: String, enumValue: MockEnumValue) {
+        self.__variables = [
+            "id": id,
+            "enumValue": GraphQLEnum(enumValue)
+        ]
+    }
+}

